Add packet attribute triage function

kimhanbeom opened the pull request:
@sehkone 
Regarding the addition of packet attribute functionality, I have added this functionality to `DnsCovertChannel` as follows. I would appreciate any feedback on this modification.


closes #354 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced functions for enhanced attribute comparison, supporting various data types and comparison scenarios.
  - Added dynamic scoring for DNS attributes, improving the accuracy of attribute evaluations.

- **Bug Fixes**
  - Updated method to log warnings for invalid DNS attribute fields, enhancing error handling.

- **Documentation**
  - Clarified method signatures for improved usability and understanding of the API.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sehkone commented on src/event/dns.rs:
```diff
@@ -3,10 +3,33 @@ use std::{fmt, net::IpAddr, num::NonZeroU8};
+};
 use crate::event::common::{triage_scores_to_string, vector_to_string};
 
+const DNS_S_IP: &str = "dns-id.orig_h";
+const DNS_S_PORT: &str = "dns-id.orig_p";
+const DNS_D_IP: &str = "dns-id.resp_h";
+const DNS_D_PORT: &str = "dns-id.resp_p";
+const DNS_PROTO: &str = "dns-proto";
+const DNS_QUERY: &str = "dns-query";
+const DNS_ANSWER: &str = "dns-answers";
+const DNS_TRANS_ID: &str = "dns-trans_id";
+const DNS_RTT: &str = "dns-rtt";
+const DNS_QCLASS: &str = "dns-qclass";
+const DNS_QTYPE: &str = "dns-qtype";
+const DNS_RCODE: &str = "dns-rcode";
+const DNS_AA_FLAG: &str = "dns-AA";
+const DNS_TC_FLAG: &str = "dns-TC";
+const DNS_RD_FLAG: &str = "dns-RD";
+const DNS_RA_FLAG: &str = "dns-RA";
+const DNS_TTL: &str = "dns-TTLs";
```
I think it would be much better to define the attributes of the enum type. Since the definition is likely to be shared across our modules, we can manage it in a new repository.

kimhanbeom commented:
@sehkone 
Are there any issues with the implementation other than moving the enum defined for the attributes to a separate repository? If not, I would like to continue with the implementation of packet attribute triage for the remaining detection events.

kimhanbeom commented:
@sehkone 

Since https://github.com/aicers/attrievent/pull/2 was merged into the main branch of the attrievent repository, I have modified the code to use the enum from that repository.
Now that we are ready to merge this PR into the review-database main branch, could you kindly review the code?

sehkone commented:
@kimhanbeom, I think it would be better to rename `PacketAttr` to `RawEventAttr`, because other raw events rather than packets are now handled.

sehkone commented on src/tables/triage_policy.rs:
```diff
@@ -135,6 +138,7 @@ impl Ord for Ti {
 
 #[derive(Clone, PartialEq, Deserialize, Serialize)]
 pub struct PacketAttr {
+    pub raw_event_type: String,
```
How about using its corresponding enum instead of a `String`?

kimhanbeom replied:
done

sehkone commented on CHANGELOG.md:
```diff
@@ -5,6 +5,35 @@ file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
+  `TorConnection`) and implementations within that module have been moved to
+  `crate::event::http`.
+- Fixed HTTP detection events to consistently use `referrer` instead of
+  `referrer` and `referrer` interchangeably.
```
I suspect there might be a typo.

kimhanbeom replied:
done

sehkone commented on src/tables/triage_policy.rs:
```diff
@@ -68,8 +68,11 @@ pub enum TiCmpKind {
 pub enum ValueKind {
     String,
-    Integer,
+    SInteger,
+    UInteger,
```
- Could you add some comments to explain the range of these values?
- If `SInteger` means `i32`, don't we need `i64`?
- If `SInteger` is `i32` and `UInteger` is `u32`, I think the pair of `Integer` and `UInteger` would be better.

kimhanbeom replied:
`SInteger` means `i64` and `UInteger` means `u64`. I'll add a comment about that.

sehkone replied:
> `SInteger` means `i64` and `UInteger` means `u64`.

In this case, I think the pair of `Integer` and `UInteger` is still better. Besides, `SInteger` even seems to mean "Small Integer".

kimhanbeom replied:
done

sehkone commented on CHANGELOG.md:
```diff
@@ -5,6 +5,35 @@ file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
+- Changed the type of fields in the detection event structure for some raw
+  event. This change allows users to see meaningful values directly without
+  having to do any special conversion for that field.
+  - `post_body`(HTTP): `Vec<u8>` to `String`.
```
If you mean `HttpThreat::post_body`, I suggest the following:
```
  - `HttpThreat::post_body`: `Vec<u8>` to `String`.
```

kimhanbeom replied:
This part represents the post_body field of all detection events of the HTTP protocol type.

sehkone replied:
```
- `post_body` inside structs of `event::http`: `Vec<u8>` to `String`.
```

How about the above for clarity?

sehkone commented on CHANGELOG.md:
```diff
@@ -5,6 +5,35 @@ file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
+  - Added a new enum type `AttrValue`. This type is used to convert the
+    attribute value of each raw event to its corresponding type to perform
+    comparison operations.
+  - Added the `target_attribute` to the `Match` trait to generate an `AttrValue`
```
I think the `target_attribute` method should be renamed to `to_attr_value`, because it literally returns `AttrValue`.

kimhanbeom replied:
done

sehkone commented on src/event/dhcp.rs:
```diff
@@ -167,9 +212,9 @@ impl BlockListDhcp {
             renewal_time: fields.renewal_time,
             rebinding_time: fields.rebinding_time,
-            class_id: fields.class_id,
+            class_id: String::from_utf8(fields.class_id.clone()).unwrap_or_default(),
```
I think the following is a better choice:

```rust
    class_id: std::str::from_utf8(&fields.class_id).unwrap_or_default().to_string(),
```

kimhanbeom replied:
done

sehkone commented on src/event/dhcp.rs:
```diff
@@ -68,7 +113,7 @@ impl fmt::Display for BlockListDhcpFields {
             self.renewal_time.to_string(),
             self.rebinding_time.to_string(),
-            to_hardware_address(&self.class_id),
+            String::from_utf8(self.class_id.clone()).unwrap_or_default(),
```
I suggest:

```rust
    std::str::from_utf8(&fields.class_id).unwrap_or_default().to_string(),
```

sehkone replied:
Since `str` has an advantage over `String` in terms of performance, and there will be virtually no cloning if `unwrap` fails, I believe this approach is slightly better.

kimhanbeom replied:
done

sehkone commented:
I think I need to explain why this new crate [attrievent](https://github.com/aicers/attrievent) has been introduced.

The kinds of raw events and their attributes change as our software evolves. The purpose of attrievent is to provide a comprehensive list of attributes for both review and the UI simultaneously.

msk commented:
It would be beneficial to break down this PR into smaller, more focused ones. The formatting fixes for old CHANGELOG entries and spelling fixes in the code are great, but they're not directly related to the main changes in this PR.

Would you mind separating those into their own PRs? That way, we can review and merge them independently, making it easier to track changes and test the main functionality of this PR. It'll also make it smaller and more manageable.

sehkone commented:
> The formatting fixes for old CHANGELOG entries and spelling fixes in the code are great, but they're not directly related to the main changes in this PR.

Regarding the changes to the CHANGELOG, I think I need to share how the team in Seoul recently adopted a method for handling text files such as CHANGELOG and README. The team is now using an extension for VS Code, Prettier, to format these files. As a result, even a small change to either file causes the entire content to be reformatted.

sehkone commented:
This needs to be rebased, updated to use the release of attrievent, and modified according to the reviewers' comments. Once that is done and [WIP] is removed, the review can proceed.

kimhanbeom commented:
> This needs to be rebased, updated to use the release of attrievent, and modified according to the reviewers' comments. Once that is done and [WIP] is removed, the review can proceed.

I've fixed it based on your comment above. Could you please review the code?

sehkone commented:
@kimhanbeom, could you rebase this?

kimhanbeom commented:
@sehkone 
I have completed the rebase.

sehkone commented on src/event.rs:
```diff
@@ -3404,7 +3403,7 @@ mod tests {
             renewal_time: 100,
             rebinding_time: 200,
-            class_id: vec![4, 5, 6],
+            class_id: "MSFT 5.0".to_string().into_bytes(),
```
I think `.as_bytes().to_vec()` would be better if it results in the same outcome.

kimhanbeom replied:
done

sehkone commented on CHANGELOG.md:
```diff
@@ -26,6 +37,18 @@ Versioning](https://semver.org/spec/v2.0.0.html).
+- Changed the type of fields in the detection event structure for some raw
+  event. This change allows users to see meaningful values directly without
+  having to do any special conversion for that field.
+  - `post_body` inside structs of `event::http`: `Vec<u8>` to `String`.
+  - `chaddr` inside structs of `event::bootp`: `Vec<u8>` to `String`.
+  - `class_id`, `client_id` inside structs of `event::dhcp`: `Vec<u8>` to
+    `String`.
```
We can't treat `Vec<u8>` and `String` as the same, because the latter supports UTF-8 characters only.

kimhanbeom replied:
done

sehkone commented on CHANGELOG.md:
```diff
@@ -14,6 +14,17 @@ Versioning](https://semver.org/spec/v2.0.0.html).
   `NonBrowser`,`ExternalDdos`, `CryptocurrencyMiningPool`, `BlockListDceRpc`
   detection events.
+- Added triage functionality for scoring with attributes of each raw event.
+  - Introduced a new crate attrievent. The kinds of raw events and their
```
I think `Used the attrievent crate` would be more accurate, because attrievent is an independent crate.

sehkone commented:
@kimhanbeom 

I think this needs rebasing. Could you do that for me finishing the review?

kimhanbeom commented:
@sehkone
I have completed the rebase.

sehkone commented on src/event/common.rs:
```diff
@@ -30,14 +32,24 @@ pub(super) trait Match {
-    fn score_by_packet_attr(&self, triage: &TriagePolicy) -> f64;
+    fn to_attr_value(&self, raw_event_attr: RawEventAttrKind) -> Option<AttrValue>;
+    fn score_by_attr(&self, packet_attr: &[PacketAttr]) -> f64 {
+        let mut total_score = 0.0;
+        for attr in packet_attr {
+            let Ok(raw_event_attr) =
+                RawEventAttrKind::from_kind_and_attr_name(&attr.raw_event_kind, &attr.attr_name)
+            else {
+                continue;
+            };
+            let Some(target_attr_val) = self.to_attr_value(raw_event_attr) else {
+                continue;
+            };
+            if process_attr_compare(target_attr_val, attr) {
+                total_score += attr.weight.unwrap_or_default();
+            }
+        }
+        (total_score * 100.0).trunc() / 100.0
+    }
```
I think this part can be refactored to use the `fold` method for more idiomatic Rust style.

sehkone replied:
(excluding the line 51)

sehkone commented on src/event/common.rs:
```diff
@@ -30,14 +32,24 @@ pub(super) trait Match {
-    /// <https://github.com/petabi/review-database/pull/321#discussion_r1721392271>
-    fn score_by_packet_attr(&self, triage: &TriagePolicy) -> f64;
+    fn to_attr_value(&self, raw_event_attr: RawEventAttrKind) -> Option<AttrValue>;
+    fn score_by_attr(&self, packet_attr: &[PacketAttr]) -> f64 {
+        let mut total_score = 0.0;
+        for attr in packet_attr {
+            let Ok(raw_event_attr) =
+                RawEventAttrKind::from_kind_and_attr_name(&attr.raw_event_kind, &attr.attr_name)
+            else {
+                continue;
+            };
+            let Some(target_attr_val) = self.to_attr_value(raw_event_attr) else {
+                continue;
+            };
+            if process_attr_compare(target_attr_val, attr) {
+                total_score += attr.weight.unwrap_or_default();
+            }
+        }
+        (total_score * 100.0).trunc() / 100.0
+    }
```
I believe this is one of the most crucial parts to understand the overall logic of this PR. Therefore, I suggest the following naming improvements to help readers better understand how it works.

```rust
    fn score_by_attr(&self, attr_triage: &[PacketAttr]) -> f64 {
        let mut total_score = 0.0;
        for item in attr_triage {
            let Ok(kind) =
                RawEventAttrKind::from_kind_and_attr_name(&item.raw_event_kind, &item.attr_name)
            else {
                continue;
            };
            let Some(value) = self.find_attr_by_kind(kind) else {
                continue;
            };
            if compare_with_attr(value, item) {
                total_score += item.weight.unwrap_or_default();
            }
        }
        (total_score * 100.0).trunc() / 100.0
    }
```

kimhanbeom replied:
done

sehkone commented on src/event/bootp.rs:
```diff
@@ -180,7 +209,7 @@ impl Match for BlockListBootp {
 
-    fn score_by_packet_attr(&self, _triage: &TriagePolicy) -> f64 {
-        0.0
+    fn to_attr_value(&self, raw_event_attr: RawEventAttrKind) -> Option<AttrValue> {
```
A `to_xxx` method typically implies that it converts the entire `struct` into a certain value, rather than returning a single field. In our case, we are selecting a field that matches a specific type, so I think the following naming would be more intuitive:

```rust
    fn find_attr_by_kind(&self, kind: RawEventAttrKind) ...
```

kimhanbeom replied:
done

sehkone commented on src/event/bootp.rs:
```diff
@@ -1,10 +1,39 @@
+use super::{EventCategory, LearningMethod, MEDIUM, TriageScore, common::Match};
+use crate::event::common::{AttrValue, to_hardware_address, triage_scores_to_string};
+
+macro_rules! bootp_target_attr {
```
Based on the alternative name `find_attr_by_kind` that I suggested for `to_attr_value`, I think it would be better to rename this accordingly.

```rust
macro_rules! find_attr_by_kind_for_bootp {
```

sehkone replied:
Could you tell me why you employ a macro not a method? I couldn't see any particular advantage.

sehkone replied:
> Could you tell me why you employ a macro not a method? I couldn't see any particular advantage.

I guess the same procedure is applied to different structures, isn't it?

kimhanbeom replied:
Yes, that's precisely the reason

kimhanbeom replied:
As per your suggestion, I have updated all macro names to `find_xxx_attr_by_kind`.

sehkone commented on src/event/common.rs:
```diff
@@ -319,15 +331,229 @@ pub fn to_hardware_address(chaddr: &[u8]) -> String {
+{
+    match packet_attr.cmp_kind {
+        AttrCmpKind::NotEqual | AttrCmpKind::NotContain => {
+            attr_val.iter().all(|x| compare_fn(x, packet_attr))
```
I guess `attr_val.iter().all(|x| !compare_fn(x, packet_attr))` might be what you intended. Could you check it out?

kimhanbeom replied:
For attributes with vector values, depending on the `AttrCmpKind`, it may be necessary to compare and satisfy all values in the vector.
For example, in the case of a `Vec<String>`, if the `AttrCmpKind` is `NotContain`, it should not match any value in the `vec<String>`, so it should be treated as `attr_val.iter().all()`.

sehkone commented on src/event/common.rs:
```diff
@@ -319,15 +331,229 @@ pub fn to_hardware_address(chaddr: &[u8]) -> String {
+    VecRaw(&'a [u8]),
+}
+
+fn check_all_or_any<'a, T, F>(attr_val: &'a [T], packet_attr: &'a PacketAttr, compare_fn: F) -> bool
```
The current name doesn’t clearly convey what the method actually does. I think the part of "all or any" is not relevant based on the context. Since this checks not just a single attribute value but multiple ones, I believe a more suitable name would be simply `is_matching_list`.

kimhanbeom replied:
done

sehkone commented on src/event/common.rs:
```diff
@@ -319,15 +331,229 @@ pub fn to_hardware_address(chaddr: &[u8]) -> String {
+    }
+}
+
+fn process_attr_compare(target_value: AttrValue, attr: &PacketAttr) -> bool {
```
I suggest this better one. Sorry for the confusion.

```rust
fn is_attr_matched(value: AttrValue, attr: ...
```

sehkone replied:
or

```rust
fn matches_attr(value: AttrValue, attr: &PacketAttr) -> bool {
```

kimhanbeom replied:
Following your suggestion, I have updated the function name to `is_attr_matched`.

sehkone commented on src/event/common.rs:
```diff
@@ -319,15 +331,229 @@ pub fn to_hardware_address(chaddr: &[u8]) -> String {
+    }
+}
+
+fn compare_bool_attribute(attr_val: bool, packet_attr: &PacketAttr) -> bool {
```
When a method returns a bool, naming it with `compare_` is not good. I suggest:

```
fn matches_bool_attr(value: bool, packet_attr: &PacketAttr) -> bool {
```

kimhanbeom replied:
As per your suggestion, I have updated all functions that compare attributes and return a boolean type to `matches_xxx_attr`.

sehkone commented:
> For the Petabi team, I think I need to explain why this new crate [attrievent](https://github.com/aicers/attrievent) has been introduced.
> 
> The kinds of raw events and their attributes change as our software evolves. The purpose of attrievent is to provide a comprehensive list of attributes for both the REview and the UI simultaneously.

I’d like to add one clarification about the attrievent. Although it provides a list of attributes, it doesn’t guarantee a complete set that covers all of them. This isn’t a significant issue, since we only use the attrievent for triaging events.

kimhanbeom commented on src/event/common.rs:
```diff
@@ -319,15 +331,229 @@ pub fn to_hardware_address(chaddr: &[u8]) -> String {
+    }
+}
+
+fn compare_all_attr_cmp_kind<T>(
```
I named it `compare_all_attr_cmp_kind` to mean that it can compare all `AttrCmpKinds`, unlike other functions (e.g. `compare_string_attribute`) that have a fixed set of comparable `AttrCmpKind` values.

kimhanbeom replied:
I have updated the function name to `matches_attr`.

sehkone commented:
@kimhanbeom,

This needs to be rebased to resolve the conflicts.

kimhanbeom commented:
pending: https://github.com/aicers/attrievent/issues/10

kimhanbeom commented:
@sehkone
I have completed the rebase.

sehkone commented:
I believe this PR completes the changes I planned to incorporate.

msk reviewed (changes requested):
I have a few observations and suggestions, mostly related to the `CHANGELOG.md` and dependency management, that I think will help with maintainability and our release process.

Regarding the `CHANGELOG.md` changes:

1.  **Formatting Commits:** I noticed quite a few formatting adjustments throughout the `CHANGELOG.md`. Generally, it's good practice to separate purely stylistic or formatting changes into their own commit (or even a separate PR). This helps keep the history clean and makes it easier to review functional changes without the noise of formatting diffs.
2.  **Formatting Rationale & Conventions:** Could you perhaps elaborate on the criteria for the formatting changes? The `review-database` project generally aims to follow `markdownlint`'s default rules, especially for line length (MD013). It wasn't immediately clear if the new formatting aligns with a specific new standard or if it was personal preference.
3.  **Changelog Content and Audience:** The `CHANGELOG.md` for `review-database` is primarily for our users – other developers who consume this crate. We try to focus on user-visible changes like new public APIs, breaking changes, significant feature additions, or bug fixes that impact behavior.
    Some of the added changelog entries, such as detailing the internal mechanics of the `Match` trait (e.g., `find_attr_by_kind`, `score_by_attr` – noting that `Match` is `pub(super)` and not part of the public crate API) or the introduction of internal enums like `AttrValue` (if it's not directly part of the public API for scoring), might be a bit too granular for the main changelog. These details are great for commit messages or the PR description itself.
    On the other hand, changes to public-facing structures like the `ValueKind` enum (which is used in `PacketAttr` for defining triage policies) are definitely important to note, and I see you've included that. We should aim to clearly communicate changes that affect how a user would interact with or configure `review-database`.

Another point is about the dependency on `attrievent`:

1.  **Dependency on Unreleased Version:** The PR currently makes `review-database` depend on a specific commit hash (`rev = "3ffc600"`) of the `attrievent` crate. Our release policy for `review-database` requires that all its dependencies are released versions. This ensures stability and predictability for our users.
2.  **Release Cycle Mismatch:** `attrievent` has a different release cadence than `review-database`. Looking at its history, `attrievent`'s last release was around December 2024, and it hasn't had a new release in the ~4.5 months since. In contrast, `review-database` has had about four releases during that same period. If we merge this PR with `review-database` depending on an unreleased `attrievent` commit, we won't be able to make any new releases of `review-database` until `attrievent` publishes a new version containing these changes. This could significantly block our ability to release updates, bug fixes, or other features for `review-database`.
3.  **Path Forward:** To avoid this, I'd strongly suggest we wait for these `attrievent` changes to be part of an official `attrievent` release. Once that's available, we can update the `Cargo.toml` here to depend on that released version. This would keep our release process smooth and unblocked.

These are mainly process and maintainability points to ensure everything integrates smoothly. The core functionality you've added looks good.

sehkone commented:
@msk, I'd like to share some context regarding formatting.

The formatting changes in this PR shouldn’t have been included in the first place.  @kimhanbeom should have turned off the Prettier setting that was recently introduced for Seoul engineers.  To avoid confusion, I had already asked @sophie-cluml and her team to disable this setting when working on `review-database`. I assume it may have been overlooked during doing rebases.

Here’s some background on the Prettier setting adopted by the Seoul engineers:

- @sophie-cluml wanted to enforce 80-character line wrapping automatically, and proposed adding `"prettier.proseWrap": "always"` to the VS Code settings.
- This option forces every line to wrap at 80 characters on save.
- The issue is that it causes unnecessary diffs in unchanged parts of the file, confusing PR reviewers.
- For each Aicers repository, we’ve been handling README reformatting separately to avoid noise.
- Specifically for `review-database`, I requested that the setting be turned off whenever developers submit PRs, as I didn’t want this project to be affected by formatting issues.

One of the reasons I didn’t want to apply this broadly was because I wasn’t confident that Prettier strictly follows Markdown conventions.  It doesn’t behave the same way as the VS Code default formatter, and it seems not to fully adhere to all markdown standards—though it clearly handles line wrapping well. Due to time constraints, I haven’t looked deeply into configuring Prettier to follow stricter Markdown rules.

@kimhanbeom, could you revert the README formatting to its previous state by disabling the setting?

sehkone commented:
@kimhanbeom, regarding the use of `attrievent`’s commit hash:

As you’re well aware, our guidelines—both documented and implied—state that release versions should be used unless there's a specific reason to make an exception. It seems this rule may have been missed during the rebase.

In addition to the Prettier setting not being turned off, this also appears to have been introduced as part of that process.  I’d recommend making it a habit to thoroughly review the code after rebasing to catch these kinds of issues early.

sehkone commented:
@msk 

I plan to have the Seoul engineers stop using Prettier for Markdown formatting, and will provide them with a detailed guideline on adhering to `markdownlint` rules.

kimhanbeom commented:
pending: https://github.com/aicers/attrievent/issues/12

kimhanbeom commented:
@sehkone @msk 
Changes based on the feedback have been applied.

kimhanbeom commented:
Added migration code for packet attribute.

- Stored attribute info
  <img width="1048" alt="image" src="https://github.com/user-attachments/assets/3910243e-b5e2-43fc-853f-7d325a8a4cc2" />
- Migrating the database to version "0.38.0-alpha.1". (using minireview)
   <img width="848" alt="image" src="https://github.com/user-attachments/assets/b068e42e-0b30-4f25-98dc-7ae209713fc5" />
- Migration result
  <img width="341" alt="image" src="https://github.com/user-attachments/assets/59084a9f-1615-49ec-870d-c742081ccd66" />

kimhanbeom commented on Cargo.toml:
```diff
@@ -51,6 +52,7 @@ tokio = { version = "1", features = ["macros"] }
 tokio-postgres-rustls = "0.13"
 toml = "0.8.13"
 tracing = "0.1"
+attrievent = { git = "https://github.com/aicers/attrievent.git", tag = "0.2.0" }
```
done

sehkone commented:
I haven’t looked into the migration part yet. I’ll share my thoughts once I’ve completed the full review.

sehkone commented:
@msk, I've finished reviewing this and I think it looks good to me.

msk commented:
@kimhanbeom,

I've noticed that this PR introduces several new macros (e.g., `find_bootp_attr_by_kind!`, `find_conn_attr_by_kind!`, etc.) which appear to be used only once in their respective modules, often within functions that seem to act primarily as shallow wrappers for these macro calls.

Could you please share the reasoning behind using macros for these cases? I'm curious to understand if there's a specific benefit or a plan for future use that I might be missing.

Given that the functions calling these macros are essentially just wrappers around the macro call, I'm wondering if inlining the macro's logic directly into those functions might be a more straightforward approach than using a macro in these instances. What are your thoughts on this?

kimhanbeom commented:
@msk 
The reason for using macros is to handle common processing through macros when different detection event structs share the same fields. (ex) `HttpThreat`, `DomainGenerationAlgorithm`) By implementing it this way, I aimed to reduce redundant code that would otherwise need to be separately implemented for each structure. Even for protocols that currently have only a single detection event defined(ex) `BlocklistBootp`), macros were used to account for possible future additions of events with the same fields.

sehkone commented:
@msk, I am just letting you know that the last pushed change only relates to updating `attrievent`.
